Rubberduck.VBEEditor/ComManagement/TypeLibs/TypeLibs.cs

```diff
@@ -0,0 +1,1669 @@
+            foreach (var knownDocClass in KnownDocumentInterfaces)
+            {
+                KnownDocumentInterfaceProgIds[index++] = knownDocClass.DocTypeInterfaceProgId;
+            }
+        }
+    }
+
+    /// <summary>
+    /// A wrapper for ITypeInfo objects, with specific extensions for VBE hosted ITypeInfos
+    /// </summary>
+    /// <remarks>
+    /// There are two significant bugs in the VBE implementations for ITypeInfo that we have to work around.
+    /// 1)  Some implementations of ITypeInfo provided by the VBE will crash with an AV if you call 
+    ///      IProvideClassInfo::GetClassInfo on them.  And guess what method the CLR calls on all COM interop objects 
+    ///      when creating a RCW?  You guessed it.  So, we use an aggregation object, plus ITypeInfo and ITypeLib wrappers 
+    ///      to circumvent this VBE bug.
+    ///
+    /// 2)  The ITypeInfo for base classes of UserForms crash with an AV if you call ITypeInfo::GetDocumentation(MEMBERID_NIL) 
+    ///     to get the type name.  We've got to remember that the VBE didn't ever intend for us to get hold of these objects, 
+    ///     so there will be little bugs.  This bug is also resolved in the provided wrappers.
+    ///
+    /// This class can also be cast to ComTypes.ITypeInfo for raw access to the underlying type information
+    /// </remarks>
+    public class TypeInfoWrapper : ComTypes.ITypeInfo, IDisposable
+    {
+        private DisposableList<TypeInfoWrapper> _typeInfosWrapped;
+        private TypeLibWrapper _containerTypeLib;
+        public TypeLibWrapper Container { get => _containerTypeLib; }
+        private int _containerTypeLibIndex;
+        private bool _isUserFormBaseClass = false;
+        private IntPtr _rawObjectPtr;
+        private ComTypes.ITypeInfo _wrappedObjectRCW;
+
+        private ComTypes.TYPEATTR _cachedAttributes;
+        public ComTypes.TYPEATTR Attributes { get => _cachedAttributes; }
+
+        private RestrictComInterfaceByAggregation<ComTypes.ITypeInfo> _ITypeInfo_Aggregator;
+        private ComTypes.ITypeInfo target_ITypeInfo { get => _ITypeInfo_Aggregator?.WrappedObject ?? _wrappedObjectRCW; }
+
+        private RestrictComInterfaceByAggregation<IVBEComponent> _IVBEComponent_Aggregator;
+        private IVBEComponent target_IVBEComponent { get => _IVBEComponent_Aggregator?.WrappedObject; }
+
+        private RestrictComInterfaceByAggregation<IVBETypeInfo> _IVBETypeInfo_Aggregator;
+        private IVBETypeInfo target_IVBETypeInfo { get => _IVBETypeInfo_Aggregator?.WrappedObject; }
+
+        public bool HasVBEExtensions { get => _IVBETypeInfo_Aggregator?.WrappedObject != null; }
+
+        private bool _hasModuleScopeCompilationErrors;
+        public bool HasModuleScopeCompilationErrors => _hasModuleScopeCompilationErrors;
+
+        /// <summary>
+        /// Exposes an enumerable collection of functions provided by the ITypeInfo
```

Comment:
Need to clarify whether this exposes all functions or only the publicly accessible functions. If all, is there a way to know the scope?

Reply:
it exposes whatever the particular ITypeInfo exposes, which at this implementation detail we can't know.    e.g. We could be wrapping a real type library provided ITypeInfo, where the behaviour is not the same as the VBE hosted ITypeInfos.    I'll add a comment to the top of the document to mention the VBE specifics